### PR TITLE
fixing FREELIST macros for AArch64

### PR DESCRIPTION
**Pull request by dmorilha**
tested on `x86_64` and `aarch64`

### CODE REVIEW

File: include/tscore/ink_queue.h [hunk 1]
```diff
@@ -137,10 +137,10 @@ union head_p {
   (_x).s.pointer = _p;                           \
   (_x).s.version = _v
 #elif defined(__x86_64__) || defined(__ia64__) || defined(__powerpc64__) || defined(__aarch64__) || defined(__mips64)
-#define FREELIST_POINTER(_x) \
-  ((void *)(((((intptr_t)(_x).data) << 16) >> 16) | (((~((((intptr_t)(_x).data) << 16 >> 63) - 1)) >> 48) << 48))) // sign extend
-#define FREELIST_VERSION(_x) (((intptr_t)(_x).data) >> 48)
-#define SET_FREELIST_POINTER_VERSION(_x, _p, _v) (_x).data = ((((intptr_t)(_p)) & 0x0000FFFFFFFFFFFFULL) | (((_v)&0xFFFFULL) << 48))
+#define FREELIST_POINTER(_x) ((void *)(((uintptr_t)(_x).data) & 0x0000FFFFFFFFFFFF))
```
[the remark this answers is not in the document]
Comment (SolidWallOfCode) [reply include/tscore/ink_queue.h]: I think I finally understand what these bit operations do - they fill bits 48..63 with the value in bit 47. I suppose the question is, does that matter? This change 0 fills those bits, which is clearly different, but is that wrong?

[the remark this answers is not in the document]
Comment (dmorilha) [reply include/tscore/ink_queue.h]: That's a fair point. I will be honest I don't have a clue why it behaves differently in `x86` and `aarch64`. My line of investigation was to instrument the code and see how that macro was affecting the pointer. In the end of the day, I think the macro should be symmetrical to `SET_FREELIST_POINTER_VERSION` - use the 48 lower bits to store the pointer - and hence my change, I would also recommend to use the unsigned versions of the types currently being used.

File: include/tscore/ink_queue.h [hunk 2]
```diff
@@ -92,8 +92,8 @@ union head_p {
   typedef int64_t version_type;
   typedef __int128_t data_type;
 #else
-  typedef int64_t version_type;
-  typedef int64_t data_type;
+  typedef uint64_t version_type;
```
[the remark this answers is not in the document]
Comment (dmorilha) [reply include/tscore/ink_queue.h]: sure thing

File: include/tscore/ink_queue.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmorilha) [reply include/tscore/ink_queue.h]: my friend @bfrancojr solved the mystery https://en.wikipedia.org/wiki/Carry_flag#Carry_flag_vs._borrow_flag

**Comment by dmorilha** [comment]
@rybit @Biilmann this is a PR you will need to run traffic server on packet.com `c1.large.arm` ;-)

**Comment by biilmann** [comment]
Very cool!

**Comment by scw00** [comment]
I haven't used x86_64 and arm64. But talked with @oknet  and @mingzym . This fix seems incorrect.

>In addition, the AMD specification requires that the most significant 16 bits of any virtual address, bits 48 through 63, must be copies of bit 47 (in a manner akin to sign extension). If this requirement is not met, the processor will raise an exception.[11](p131) Addresses complying with this rule are referred to as "canonical form."[11](p130) Canonical form addresses run from 0 through 00007FFF'FFFFFFFF, and from FFFF8000'00000000 through FFFFFFFF'FFFFFFFF, for a total of 256 TB of usable virtual address space. This is still 65,536 times larger than the virtual 4 GB address space of 32-bit machines.

https://en.wikipedia.org/wiki/X86-64

You might need to expand the sign extension to 48:63 bits.

**Comment by oknet** [comment]
According to the https://www.kernel.org/doc/Documentation/arm64/memory.txt

The one of the valid VA range for AArch64 is:

- user: 0000000000000000 ~ 0000ffffffffffff
- kernel: ffff000000000000 ~ ffffffffffffffff

```
AArch64 Linux memory layout with 4KB pages + 4 levels:

Start			End			Size		Use
-----------------------------------------------------------------------
0000000000000000	0000ffffffffffff	 256TB		user
ffff000000000000	ffffffffffffffff	 256TB		kernel
```

And the valid range for AMD64 is:

- user: 0000000000000000 ~ 00007fffffffffff
- kernel: ffff800000000000 ~ ffffffffffffffff

**Comment by oknet** [comment]
It is better to revert the commit 85634092e22365b33a64655fb479be6b2ac8639b

and fix "#define FREELIST_VERSION(_x) ((((intptr_t)(_x).data)<<1)>>49)" bug.

```
#elif defined(__x86_64__)
#define FREELIST_POINTER(_x) ((void*)(((((intptr_t)(_x).data)>>63)<<48)|(((intptr_t)(_x).data)&0xFFFFFFFFFFFFLL)))
#define FREELIST_VERSION(_x) ((((intptr_t)(_x).data)&0x7FFF000000000000LL)>>48)
#define SET_FREELIST_POINTER_VERSION(_x,_p,_v) \
  (_x).data = ((((intptr_t)(_p))&0x8000FFFFFFFFFFFFLL) | (((_v)&0x7FFFLL) << 48))
#else
#error "unsupported processor"
#endif
```

The original design (before 85634092e22365b33a64655fb479be6b2ac8639b),

- 0 ~ 47 bit stores the VA address
- 48 ~ 62 bit stores the version
- 63 bit indicator the VA address is kernel space or user space


The `SET_FREELIST_POINTER_VERSION(_x,_p,_v)` store VA(_p) and VERSION(_v) into FREELIST_POINTER(_x):

- Each bit of High16(_p) has the same value, so it only needs to save 1 of the 16 bits, the highest bit.
- It save the highest 1 bit and the lowest 48 bits by `(((intptr_t)(_ p))) and 0x8000FFFFFFFFFFFFLL)`.
- And the rest 15 of 16 bits is used to save VERSION(_v) by `| (((_v)&0x7FFFLL) << 48)`

The `FREELIST_POINTER(_x)` return VA from a FREELIST_POINTER:

- if the VA address is from kernel space (start with 0xFFFFxxxxxxxxxxxx), the `FREELIST_POINTER(_x)` have to do 0xFFFF000000000000 | Low48(_x) and return, or just return Low48(_x).
- Low48(_x) = `(((intptr_t)(_x).data)&0xFFFFFFFFFFFFLL)` 
- 0xFFFF000000000000 or 0x0000000000000000 = `((((intptr_t)(_x).data)>>63)<<48)`
   - It fills the sign bit into all bits by right shift 63 times.
   - And clear the low 48 bits by left shift 48 times.

The `FREELIST_VERSION(_x)` return VERSION from a FREELIST_POINTER, but there is a slight bug with `#define FREELIST_VERSION(_x) ((((intptr_t)(_x).data)<<1)>>49)`

- It tries to clear the highest bit by left shift 1 time, but it causes the sign bit filled by next bit.
- And then get the 15 bits VERSION by right shift 49 times.

We can just fix it simply by `#define FREELIST_VERSION(_x) ((((intptr_t)(_x).data)&0x7FFF000000000000LL)>>48)`:

- Just get 15 bits by `(((intptr_t)(_x).data)&0x7FFF000000000000LL)`
- And then get the 15 bits VERSION by right shift 48 times.

**Comment by linnaea** [comment]
I don't see a point in being "correct" as defined by the language standard for this case, as we're already assuming the size of the address space and mangling pointers.

I think it's easier to just assume the top 16 bits are always 0, and if we want to be sure about it, add a check in the macro and abort early if that assumption does not hold.

**Comment by scw00** [comment]
close via #5451

**Pull request closed without merging**